File: tests/containers.cpp
#include "../lib/ytest/ytest.hpp"
#include "../luabinds.hpp"
#include <iostream>

static const std::string Testfile = "tests/lua/containers.lua";

yTestPackage containers([]{
    describe("container handling", []{
        std::vector<int> data{1,2,3};
        using data_value_type = decltype(data)::value_type;

        lua::Lua lua;
        lua.attachWithState("v_size", [data = &data](lua::State& s) { s.push(data->size()); });
        lua.attachWithState("v_clear", [data = &data](lua::State&) { data->clear(); });
        lua.attachWithState("v_empty", [data = &data](lua::State& s) { s.push(data->empty()); });
        lua.attachWithState("v_at", [data = &data](lua::State& s) {
            for (int i = 1; i <= s.sizeParameters(); ++i) {
                s.push(data->at(s.get<std::size_t>(i)));
            }
        });
        lua.attachWithState("v_set", [data = &data](lua::State& s) {
            data->at(s.get<int>(1)) = s.get<data_value_type>(2);
        });
        lua.attachWithState("v_push", [data = &data](lua::State& s) {
            data->push_back(s.get<data_value_type>(1));
        });

        lua.eval(Testfile);

        it("it can cope with containers", [&]{
            Assert().isFalse(lua.get<bool>("fail"));
        });
    });
});

File: tests/lua/containers.lua
fail = false

if (v_size() ~= 3 or v_empty() ~= false) then fail = true end
if (v_at(1) ~= 2 or v_at(2) ~= 3) then fail = true end

v_set(1, 11)
if (v_at(1) ~= 11) then fail = true end

v_clear()
if (v_size() ~= 0 or v_empty() ~= true) then fail = true end

v_push(42)
if (v_size() ~= 1 or v_empty() ~= false or v_at(0) ~= 42) then fail = true end
